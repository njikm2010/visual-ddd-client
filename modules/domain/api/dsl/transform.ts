/**
 * DSL 转换
 */
export {};
