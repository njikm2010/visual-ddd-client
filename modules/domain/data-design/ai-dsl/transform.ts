import { DataObjectDSL, DataObjectTypeDSL, DataObjectTypeName, DataObjectPropertyDSL } from '../dsl/dsl';
import { NameDSL } from '../../domain-design/dsl/dsl';
import { UntitledInHumanReadable } from '../../domain-design/dsl/constants';
import snakeCase from 'lodash/snakeCase';
import { assert } from '@/lib/utils';

function getTableName(table: DataObjectDSL) {
  return table.tableName || snakeCase(table.name);
}

function getPropertyName(property: DataObjectPropertyDSL) {
  return property.propertyName || snakeCase(property.name);
}

function stringifyType(type: DataObjectTypeDSL, tables: DataObjectDSL[]) {
  if (type.type !== DataObjectTypeName.Reference) {
    return type.type;
  }

  const refTable = tables.find(i => i.uuid === type.target);

  assert(refTable, `未找到引用的数据表 ${type.target}`);

  const refProperty = refTable.properties.find(i => i.uuid === type.targetProperty);

  assert(refProperty, `未找到引用的数据表 ${refTable.name} 的属性 ${type.targetProperty}`);

  return `${type.type}(${getTableName(refTable)}.${getPropertyName(refProperty)})`;
}

function getComment(data: NameDSL) {
  if (data.title && !data.title.includes(UntitledInHumanReadable)) {
    return data.title;
  }
  return undefined;
}

function stringifyProperty(property: DataObjectPropertyDSL, table: DataObjectDSL, tables: DataObjectDSL[]): string {
  const type = stringifyType(property.type, tables);
  const modifier: string[] = [];

  if (property.primaryKey) {
    modifier.push('PrimaryKey');
  } else if (property.notNull) {
    modifier.push('NotNull');
  }

  // TODO: 支持默认值
  const comment = getComment(property);

  return `  ${getPropertyName(property)}: ${type}${modifier.length ? ', ' : ''}${modifier.join(', ')};${
    comment ? ` #${comment}` : ''
  }`;
}

function stringifyTable(table: DataObjectDSL, tables: DataObjectDSL[]): string {
  const comment = getComment(table);
  return `${comment ? `#${comment}\n` : ''}Table ${getTableName(table)} (\n${table.properties
    .map(i => {
      return stringifyProperty(i, table, tables);
    })
    .join('\n')}
)`;
}

function stringifyIndex(table: DataObjectDSL): string {
  if (table.indexes.length) {
    let dsl = '';
    for (const index of table.indexes) {
      if (!index.properties.length) {
        continue;
      }

      const idx = `Index ${getTableName(table)}.${index.name} (
	type: ${index.type};
	method: ${index.method};
	fields: ${index.properties
    .map(p => {
      const property = table.properties.find(i => i.uuid === p);
      assert(property, `未找到属性 ${p}`);

      return getPropertyName(property);
    })
    .join(', ')};
)`;
      dsl += idx + '\n\n';
    }

    return dsl;
  }

  return '';
}

/**
 * 将数据表 DSL 转换为 AI 容易识别的，简洁的格式
 */
export function transform(tables: DataObjectDSL[], allTables: DataObjectDSL[]): string {
  let dsl = '';

  for (const table of tables) {
    const tbl = stringifyTable(table, allTables);
    const index = stringifyIndex(table);

    dsl += `${tbl}\n\n${index}`;
  }

  return dsl;
}
