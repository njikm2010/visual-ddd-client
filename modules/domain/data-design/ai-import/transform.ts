import { AIImport } from './dsl';
import { DataObjectDSL, DataObjectReference, DataObjectTypeName } from '../dsl/dsl';
import { createDataObjectDSL, createDataObjectPropertyDSL, createDataObjectType } from '../dsl/factory';
import { assert } from '@/lib/utils';

function isReference(
  type: AIImport.DataObjectNormalType | AIImport.DataObjectReferenceType
): type is AIImport.DataObjectReferenceType {
  return type.type === DataObjectTypeName.Reference;
}

export function transform(data: AIImport.DataObject[]): DataObjectDSL[] {
  const pending: Function[] = [];
  const list: DataObjectDSL[] = [];

  for (const item of data) {
    const object = createDataObjectDSL();
    object.title = item.title;
    object.name = item.name;
    object.properties = [];

    for (const prop of item.properties) {
      const property = createDataObjectPropertyDSL();
      property.name = prop.name;
      property.title = prop.title;

      if (prop.primaryKey) {
        property.notNull = true;
        property.primaryKey = true;
      }

      property.type = createDataObjectType(prop.type.type ?? DataObjectTypeName.String);

      if (isReference(prop.type)) {
        const type = prop.type;
        pending.push(() => {
          const obj = list.find(i => i.name === type.target);
          assert(obj, `引用类型解析错误, 未找到对象 ${type.target}`);

          const field = obj.properties.find(i => i.name === type.property);

          assert(field, `引用类型解析错误, 未找到字段 ${type.target}.${type.property}`);

          const ref = property.type as DataObjectReference;
          ref.target = obj.uuid;
          ref.targetProperty = field.uuid;
          ref.cardinality = type.cardinality;
        });
      }

      object.properties.push(property);
    }

    list.push(object);
  }

  for (const fn of pending) {
    fn();
  }

  return list;
}
