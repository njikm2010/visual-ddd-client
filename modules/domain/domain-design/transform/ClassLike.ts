import { createEntityIdProperty } from '../dsl';
import type { ClassDSL, DTODSL, EntityDSL, ValueObjectDSL } from '../dsl/dsl';

import { PropertiesLike } from './PropertiesLike';

export class ClassLike<T extends ClassDSL> extends PropertiesLike<T> {
  override clone() {
    const clone = super.clone();

    // 拷贝类特有的属性
    this.regenerateUUIDList(clone.classProperties);
    this.regenerateUUIDList(clone.methods);
    this.regenerateUUIDList(clone.classMethods);

    return clone;
  }

  override toValueObject(): ValueObjectDSL {
    return this.clone();
  }

  override toDTO(): DTODSL {
    return this.clone();
  }

  override toEntity(): EntityDSL {
    const entity = this.clone() as unknown as EntityDSL;

    entity.isAggregationRoot = false;
    const id = this.getEntityId();
    if (id) {
      entity.id = id;
    } else {
      const idProperty = createEntityIdProperty();
      entity.id = idProperty.uuid;
      entity.properties.unshift(idProperty);
    }

    return entity;
  }
}
