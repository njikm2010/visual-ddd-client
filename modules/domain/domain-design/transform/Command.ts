import cloneDeep from 'lodash/cloneDeep';
import { CommandDSL, PropertiesLikeDSL } from '../dsl';
import { PropertiesLike } from './PropertiesLike';

export class Command extends PropertiesLike<CommandDSL> {
  override clone() {
    const clone = super.clone();
    this.regenerateUUIDList(clone.eventProperties);

    return clone;
  }

  override toCommand() {
    return this.clone();
  }

  protected override mergeBaseInfo(target: PropertiesLikeDSL): PropertiesLikeDSL {
    const result = super.mergeBaseInfo(target);

    if (this.current.eventProperties.length) {
      // 合并事件属性
      for (const p of this.current.eventProperties) {
        result.properties.push(this.regenerateUUID(cloneDeep(p)));
      }
    }

    return result;
  }
}
