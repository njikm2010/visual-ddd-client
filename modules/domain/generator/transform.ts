import { cloneDeep } from '@wakeapp/utils';
import { TypeDSL, TypeType, ReferenceTypeDSL, ContainerTypeDSL } from '../domain-design/dsl/dsl';
import { DomainObjectName } from '../domain-design/dsl/constants';
import { DataObjectTypeDSL, DataObjectTypeName } from '../data-design/dsl/dsl';
import { BaseTypeCompatibleMap } from '../mapper-design/dsl/compatible';

import { IDomainObjectStore } from './types';

export function transformTypeDSLReference(type: TypeDSL, transform: (type: ReferenceTypeDSL) => TypeDSL): TypeDSL {
  switch (type.type) {
    case TypeType.Base:
      return cloneDeep(type);
    case TypeType.Container:
      return {
        ...type,
        params: Object.keys(type.params).reduce((acc, key) => {
          const currentValue = type.params[key];
          acc[key] = currentValue && transformTypeDSLReference(currentValue, transform);
          return acc;
        }, {} as ContainerTypeDSL['params']),
      };
    case TypeType.Reference:
      return transform(type);
  }
}

/**
 * 转换 领域对象类型 到 数据模型类型
 * 目前仅支持基础类型 和 枚举
 * @param type
 * @returns
 */
export function transformTypeDSLToDataObjectTypeDSL(
  type: TypeDSL,
  getObjectById?: IDomainObjectStore['getObjectById']
): DataObjectTypeDSL | undefined {
  switch (type.type) {
    case TypeType.Base:
      const compatibleType = BaseTypeCompatibleMap[type.name][0] || DataObjectTypeName.String;

      return {
        type: compatibleType,
      };
    case TypeType.Reference: {
      const obj = getObjectById?.(type.referenceId);

      if (obj?.type === DomainObjectName.Enum) {
        return {
          type: obj.value.baseType === 'number' ? DataObjectTypeName.Integer : DataObjectTypeName.String,
        };
      }
    }
    default:
      return undefined;
  }
}
