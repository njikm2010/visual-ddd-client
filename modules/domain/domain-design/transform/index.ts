import { CommandDSL, DTODSL, EntityDSL, NameDSL, QueryDSL, ValueObjectDSL } from '../dsl/dsl';
import { DomainObjectName } from '../dsl/constants';

import { Command } from './Command';
import { Query } from './Query';
import { DTO } from './DTO';
import { Entity } from './Entity';
import { ValueObject } from './ValueObject';
import { IDomainObjectTransform } from './Transform';

export type { IDomainObjectTransform } from './Transform';

/**
 * 创建领域对象转换器
 * @param type
 * @param value
 */
export function createDomainObjectTransform(type: DomainObjectName.Command, value: CommandDSL): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName.Query, value: QueryDSL): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName.DTO, value: DTODSL): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName.Entity, value: EntityDSL): IDomainObjectTransform;
export function createDomainObjectTransform(
  type: DomainObjectName.ValueObject,
  value: ValueObjectDSL
): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName, value: NameDSL): IDomainObjectTransform {
  switch (type) {
    case DomainObjectName.Command:
      return new Command(value as CommandDSL);
    case DomainObjectName.Query:
      return new Query(value as QueryDSL);
    case DomainObjectName.DTO:
      return new DTO(value as DTODSL);
    case DomainObjectName.Entity:
      return new Entity(value as EntityDSL);
    case DomainObjectName.ValueObject:
      return new ValueObject(value as ValueObjectDSL);
    default:
      throw new Error(`不支持的领域对象类型: ${type}`);
  }
}

export const TRANSFORM_TARGET: [DomainObjectName, keyof IDomainObjectTransform][] = [
  [DomainObjectName.Command, 'toCommand'],
  [DomainObjectName.Query, 'toQuery'],
  [DomainObjectName.DTO, 'toDTO'],
  [DomainObjectName.Entity, 'toEntity'],
  [DomainObjectName.ValueObject, 'toValueObject'],
];
