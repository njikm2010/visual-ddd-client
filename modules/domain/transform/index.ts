import type { EditorContextMenuItem, CopyPayload } from '@/lib/editor';
import type { Node } from '@antv/x6';

import { CommandDSL, DTODSL, EntityDSL, NameDSL, QueryDSL, ValueObjectDSL } from '../domain-design/dsl/dsl';
import { DomainObjectName, DomainObjectReadableName } from '../domain-design/dsl/constants';
import { DataObjectName, DataObjectReadableName } from '../data-design/dsl/constants';

import { Command } from './Command';
import { Query } from './Query';
import { DTO } from './DTO';
import { Entity } from './Entity';
import { ValueObject } from './ValueObject';
import { IDomainObjectTransform } from './Transform';

export type { IDomainObjectTransform } from './Transform';

export type DomainObjectTransformable =
  | DomainObjectName.Command
  | DomainObjectName.Query
  | DomainObjectName.DTO
  | DomainObjectName.Entity
  | DomainObjectName.ValueObject;

export type DomainObjectTransformTarget = DomainObjectTransformable | DataObjectName.DataObject;

/**
 * 创建领域对象转换器
 * @param type
 * @param value
 */
export function createDomainObjectTransform(type: DomainObjectName.Command, value: CommandDSL): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName.Query, value: QueryDSL): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName.DTO, value: DTODSL): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectName.Entity, value: EntityDSL): IDomainObjectTransform;
export function createDomainObjectTransform(
  type: DomainObjectName.ValueObject,
  value: ValueObjectDSL
): IDomainObjectTransform;
export function createDomainObjectTransform(type: DomainObjectTransformable, value: NameDSL): IDomainObjectTransform {
  switch (type) {
    case DomainObjectName.Command:
      return new Command(value as CommandDSL);
    case DomainObjectName.Query:
      return new Query(value as QueryDSL);
    case DomainObjectName.DTO:
      return new DTO(value as DTODSL);
    case DomainObjectName.Entity:
      return new Entity(value as EntityDSL);
    case DomainObjectName.ValueObject:
      return new ValueObject(value as ValueObjectDSL);
    default:
      throw new Error(`不支持的领域对象类型: ${type}`);
  }
}

export const TRANSFORM_TARGET: [DomainObjectTransformTarget, keyof IDomainObjectTransform][] = [
  [DomainObjectName.Command, 'toCommand'],
  [DomainObjectName.Query, 'toQuery'],
  [DomainObjectName.DTO, 'toDTO'],
  [DomainObjectName.Entity, 'toEntity'],
  [DomainObjectName.ValueObject, 'toValueObject'],
  [DataObjectName.DataObject, 'toDataObject'],
];

const DOMAIN_OBJECT_TRANSFORM_TARGET_READABLE_NAME: Record<DomainObjectTransformTarget, string> = {
  ...DomainObjectReadableName,
  ...DataObjectReadableName,
};

/**
 * 创建右键菜单
 * @param type
 * @returns
 */
export function createCopyAsMenu(type: DomainObjectTransformable): EditorContextMenuItem {
  return {
    key: 'transform',
    label: '复制为',
    children: TRANSFORM_TARGET.map(([name, method]) => {
      return {
        key: name,
        label: DOMAIN_OBJECT_TRANSFORM_TARGET_READABLE_NAME[name],
        handler: context => {
          const { model, cell } = context.target!;
          const transform = createDomainObjectTransform(type as any, model.properties as any);
          const dsl = transform[method]();

          const payload: CopyPayload = {
            id: dsl.uuid,
            type: 'node',
            size: (cell as Node).getSize(),
            position: { x: 0, y: 0 },
            properties: {
              ...dsl,
              __node_name__: name,
              __node_type__: 'node',
            },
          };

          context.canvasModel.handleCopyPayloads([payload]);
        },
      };
    }),
  };
}
