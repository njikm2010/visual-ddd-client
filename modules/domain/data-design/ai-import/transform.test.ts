import { transform } from './transform';
import * as mockedUuid from 'uuid';

jest.mock('uuid', () => {
  let count = 0;
  return {
    __esModule: true,
    reset() {
      count = 0;
    },
    v4() {
      return `uuid-${count++}`;
    },
  };
});

const mockedUuidModule = mockedUuid as unknown as jest.Mocked<{ reset(): void }>;

beforeEach(() => {
  mockedUuidModule.reset();
});

test('transform', () => {
  expect(
    transform([
      {
        name: 'User',
        title: '用户',
        properties: [
          { name: 'id', title: '用户唯一id', primaryKey: true, type: { type: 'Long' } },
          { name: 'name', title: '用户名', type: { type: 'String' } },
        ],
      },
      {
        name: 'Asset',
        title: '资产',
        properties: [
          { name: 'id', title: '唯一id', primaryKey: true, type: { type: 'Long' } },
          { name: 'name', title: '资产名称', type: { type: 'String' } },
          { name: 'value', title: '资产价值', type: { type: 'Decimal' } },
          {
            name: 'userId',
            title: '用户引用',
            type: { type: 'Reference', target: 'User', property: 'id', cardinality: 'OneToMany' },
          },
        ],
      },
    ] as any)
  ).toMatchSnapshot();

  expect(() => {
    transform([
      {
        name: 'User',
        title: '用户',
        properties: [
          { name: 'id', title: '用户唯一id', primaryKey: true, type: { type: 'Long' } },
          { name: 'name', title: '用户名', type: { type: 'String' } },
        ],
      },
      {
        name: 'Asset',
        title: '资产',
        properties: [
          { name: 'id', title: '唯一id', primaryKey: true, type: { type: 'Long' } },
          { name: 'name', title: '资产名称', type: { type: 'String' } },
          { name: 'value', title: '资产价值', type: { type: 'Decimal' } },
          {
            name: 'userId',
            title: '用户引用',
            type: { type: 'Reference', target: 'User', property: 'unknown', cardinality: 'OneToMany' },
          },
        ],
      },
    ] as any);
  }).toThrowError('引用类型解析错误, 未找到字段 User.unknown');
});
