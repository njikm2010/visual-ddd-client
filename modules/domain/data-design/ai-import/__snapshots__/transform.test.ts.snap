// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform 1`] = `
[
  {
    "description": "",
    "indexes": [],
    "meta": [],
    "name": "User",
    "properties": [
      {
        "description": "",
        "meta": [],
        "name": "id",
        "notNull": true,
        "primaryKey": true,
        "propertyName": undefined,
        "title": "用户唯一id",
        "type": {
          "autoIncrement": false,
          "defaultValue": undefined,
          "type": "Long",
        },
        "uuid": "uuid-2",
      },
      {
        "description": "",
        "meta": [],
        "name": "name",
        "notNull": false,
        "primaryKey": false,
        "propertyName": undefined,
        "title": "用户名",
        "type": {
          "defaultValue": undefined,
          "length": undefined,
          "type": "String",
        },
        "uuid": "uuid-3",
      },
    ],
    "tableName": undefined,
    "title": "用户",
    "uuid": "uuid-0",
  },
  {
    "description": "",
    "indexes": [],
    "meta": [],
    "name": "Asset",
    "properties": [
      {
        "description": "",
        "meta": [],
        "name": "id",
        "notNull": true,
        "primaryKey": true,
        "propertyName": undefined,
        "title": "唯一id",
        "type": {
          "autoIncrement": false,
          "defaultValue": undefined,
          "type": "Long",
        },
        "uuid": "uuid-6",
      },
      {
        "description": "",
        "meta": [],
        "name": "name",
        "notNull": false,
        "primaryKey": false,
        "propertyName": undefined,
        "title": "资产名称",
        "type": {
          "defaultValue": undefined,
          "length": undefined,
          "type": "String",
        },
        "uuid": "uuid-7",
      },
      {
        "description": "",
        "meta": [],
        "name": "value",
        "notNull": false,
        "primaryKey": false,
        "propertyName": undefined,
        "title": "资产价值",
        "type": {
          "defaultValue": undefined,
          "precision": undefined,
          "scale": undefined,
          "type": "Decimal",
        },
        "uuid": "uuid-8",
      },
      {
        "description": "",
        "meta": [],
        "name": "userId",
        "notNull": false,
        "primaryKey": false,
        "propertyName": undefined,
        "title": "用户引用",
        "type": {
          "cardinality": "OneToMany",
          "target": "uuid-0",
          "targetProperty": "uuid-2",
          "type": "Reference",
        },
        "uuid": "uuid-9",
      },
    ],
    "tableName": undefined,
    "title": "资产",
    "uuid": "uuid-4",
  },
]
`;
